rts/motoko-rts/src/memory.rs

```diff
@@ -83,3 +83,42 @@ pub unsafe fn alloc_array<M: Memory>(mem: &mut M, tag: Tag, len: usize) -> Value
 
     skewed_ptr
 }
+
+/// Allocate a new weak reference.
+#[enhanced_orthogonal_persistence]
+#[ic_mem_fn]
+pub unsafe fn alloc_weak_ref<M: Memory>(mem: &mut M, target: Value) -> Value {
+    use crate::barriers::allocation_barrier;
+
+    let weak_ref = mem.alloc_words(size_of::<WeakRef>() + Words(16));
+
+    let weak_ref_obj = weak_ref.get_ptr() as *mut WeakRef;
+    (*weak_ref_obj).header.tag = TAG_WEAK_REF;
+    (*weak_ref_obj).header.init_forward(weak_ref);
+    (*weak_ref_obj).field = target;
```

Comment:
Here a write barrier is needed:

```suggestion
    init_with_barrier(mem, &mut (*weak_ref_obj).field, target);
```